modules/dists/SparseBlockDist.chpl

```diff
@@ -117,6 +118,82 @@ class SparseBlockDom: BaseSparseDom {
     dsiAdd((ind,));
   }
 
+  proc dsiBulkAdd(inds: [] index(rank, idxType), isSorted=false, isUnique=false,
+      preserveInds=true){
+  
+    if !isSorted && preserveInds {
+      var _inds = inds;
+      bulkAdd_help(_inds, isSorted, isUnique); 
+    }
+    else {
+      bulkAdd_help(inds, isSorted, isUnique);
+    }
+  }
+
+  // Tried to put this record in the function and the if statement, but got a
+  // segfault from the compiler.
+  record TargetLocaleComparator {
+
+    // I wish I could return tuple from key function. There are checks to see
+    // if the type of the function is numeric/string. Maybe there should be
+    // isComparable that can do recursive checks for at least tuples of
+    // numericals/strings/comparables.
+    // In general being able to return tuples from key() can enable sorting with
+    // multiple keys.
+
+    //proc key(a: index(rank, idxType)) { 
+    //  return dist.targetLocales[dist.targetLocsIdx(a)].id; 
+    //}
+
+    proc compare(a: index(rank, idxType), b: index(rank, idxType)) {
+      var locDif = dist.targetLocales[dist.targetLocsIdx(a)].id -
+        dist .targetLocales[dist.targetLocsIdx(b)].id;
+
+      if locDif != 0 then return locDif;
+
+      if a < b then return -1;
+      if a > b then return 1;
+      return 0;
+    }
+  }
+
+  proc bulkAdd_help(inds: [] index(rank,idxType), isSorted=false, isUnique=false) {
+
+    if !isSorted {
+
+      // without _new_ record functions throw null deref. It doesn't seem to be 
+      // able to deref _outer_ ? I think it has something to do with memory
+      // allocation for classes vs records, but I cannot explain
+      var comp = new TargetLocaleComparator;
+      sort(inds, comparator=comp);
+    }
+
+    var active : atomic int;
+    var firstIndex = inds.domain.low;
+    var curLoc = dist.targetLocsIdx(inds[inds.domain.low]);
+
+    for i in inds.domain {
+      const _tmpLoc = dist.targetLocsIdx(inds[i]);
+      if _tmpLoc != curLoc {
+        spawnBulkAdd(firstIndex..i-1, curLoc);
+        curLoc = _tmpLoc;
+        firstIndex = i;
+      }
+    }
+    spawnBulkAdd(firstIndex..inds.domain.high, curLoc);
+
+    active.waitFor(0);
+
+    proc spawnBulkAdd(indsRange: range, loc){
+      active.add(1);
+      begin on dist.targetLocales(loc) {
```

Comment:
@noakesmichael clarified many questions I had in mind and explained what the compiler does currently. I have to think about some of the things he said, and change the code/add notes accordingly.